lazyproto/decode_test.go

```diff
@@ -0,0 +1,1728 @@
+	"fmt"
+	"math"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/CrowdStrike/csproto"
+)
+
+func ExampleDecodeResult_FieldData() {
+	// data contains the serialized bytes of an Example Protobuf message that is
+	// defined as:
+	//	message Example {
+	//		int64  id     = 1;
+	//		Nested nested = 2;
+	//	}
+	//	message Nested {
+	//		string name        = 1;
+	//		string description = 2;
+	//	}
+	data := []byte{
+		// id: 1
+		(1 << 3), 0x1,
+		// nested: 10 bytes
+		(2 << 3) | 2, 0x0A,
+		// nested.name: foo
+		(1 << 3) | 2, 0x03, 0x66, 0x6f, 0x6f,
+		// nested.description: bar
+		(2 << 3) | 2, 0x03, 0x62, 0x61, 0x72,
+	}
+	def := NewDef()
+	// extract tags 1 and 2 from the nested message at tag 2 in the outer message
+	_ = def.AddNested(2, 1, 2)
+	res, err := Decode(data, def)
+	if err != nil {
+		fmt.Println("error from decode:", err)
+		return
+	}
+	// grab the field data
+	nameData, err := res.FieldData(2, 1)
+	if err != nil {
+		fmt.Println("error accessing field data for 'name':", err)
+		return
+	}
+	descriptionData, err := res.FieldData(2, 2)
+	if err != nil {
+		fmt.Println("error accessing field data for 'description':", err)
+		return
+	}
+	// extract the values
+	name, err := nameData.StringValue()
+	if err != nil {
+		fmt.Println("error retrieving string value for 'name':", err)
+		return
+	}
+	description, err := descriptionData.StringValue()
+	if err != nil {
+		fmt.Println("error retrieving string value for 'description':", err)
+		return
+	}
+	fmt.Printf("name: %s\ndescription: %s", name, description)
+	// Output:
+	// name: foo
+	// description: bar
+}
+
+func TestDecodePartial(t *testing.T) {
+	var sampleMessage = []byte{
+		// field 1: varint boolean true
+		(1 << 3), 0x01,
+		// field 2: string "testing"
+		(2 << 3) | 2, 0x07, 0x74, 0x65, 0x73, 0x74, 0x69, 0x6e, 0x67,
+		// field 3: nested message (10 bytes)
+		// . field 1: integer 5
+		// . field 2: string "nested"
+		(3 << 3) | 2, 0x0a, (1 << 3), 0x05, (2<<3 | 2), 0x06, 0x6e, 0x65, 0x73, 0x74, 0x65, 0x64,
+		// field 4: fixed32 1138
+		(4 << 3) | 5, 0x72, 0x04, 0x00, 0x00,
+		// field 5: float64 42.1138
+		(5 << 3) | 1, 0x74, 0x24, 0x97, 0xFF, 0x90, 0x0E, 0x45, 0x40,
+	}
+	t.Parallel()
+	t.Run("decode empty buffer", func(t *testing.T) {
+		t.Parallel()
+		res, err := Decode([]byte{}, NewDef(1))
+
+		assert.NoError(t, err)
+		assert.Empty(t, res.m)
+	})
+	t.Run("decode with nil def", func(t *testing.T) {
+		t.Parallel()
+		res, err := Decode([]byte{0x01}, nil)
+
+		assert.NoError(t, err)
+		assert.Empty(t, res.m)
+	})
+	t.Run("decode with empty def", func(t *testing.T) {
+		t.Parallel()
+		res, err := Decode([]byte{0x01}, NewDef())
+
+		assert.NoError(t, err)
+		assert.Empty(t, res.m)
+	})
+	t.Run("decode with missing def keys", func(t *testing.T) {
+		t.Parallel()
+		def := NewDef(42, 100)
+		res, err := Decode(sampleMessage, def)
+
+		assert.NoError(t, err)
+		assert.Empty(t, res.m)
+	})
+	t.Run("decode with matching def keys", func(t *testing.T) {
```

Comment:
food for though: I like adding a `should fail` or `should pass` to the test case for additional context so I don't have to force myself to reread the implementation of the test to get the gist of it. Old TDD days are hard to shake 😅. I do think it adds value though. Definitely not a blocker